### DISCUSSION TITLE
Smooth mesh exported from FreeCAD

### DISCUSSION THREAD

**Discussion by buddhasystem** [mitsuba-renderer/mitsuba3#1277] [Show and tell]
I need to model a spherical mirror, and I'm using FreeCAD for that. When I use the default export to OBJ, so that I can build a Mitsuba scene. It turns out that the default mesh in FreeCAD is rather crude, and although the object still looks smooth to the eye, the reflection from such mirror is really granular. To properly mesh objects in FreeCAD one needs to use the Mesh workbench and adjust the meshing parameters as needed. With the precision on the micron scale, and around 34k faces, I get a much better result as shown below. The emitter is "spotlight" and it's shooting from the observer towards the mirror, which then focuses the light downwards and to the side.

![image](https://github.com/user-attachments/assets/dd5dc1b2-273e-452f-86ba-e1763d6b6aaf)

**Comment by merlinND** [comment]
Hello @buddhasystem,

Thanks for sharing your finding.
I wonder if such high-resolution geometry is needed though, if the normals could be interpolated instead. Is there an option equivalent to [Blender's "Shade Smooth"](https://docs.blender.org/manual/en/2.91/scene_layout/object/editing/shading.html) in FreeCAD?
Otherwise, you could try *not* exporting the normals with your mesh and letting Mitsuba recalculate them. I think they will be smoothly interpolated by default (not sure).

**Reply by buddhasystem** [reply]
@merlinND 

Thank you so much for the note. My team needs quality mirror models (and the quality needs to be quantified of course, which I will do later) because we work with images produced in physics processes and base our reconstruction on what we detect. So the actual result depends on whether you distort let's say a ring or a spot -- again, I need to understand the acceptable distortion level.

Regarding Blender -- and thank you for reference -- , it is clear from documentation that the mesh itself remains same (coarse) but during the shading process there is smoothing out in real time. This doesn't map on the functionality of FreeCAD as far as I can tell.

Regarding the normals, it seems there is no way to disable the normals in the FreeCAD export. But, I can remove the "vn" lines from the OBJ file using a simple script/filter. In that case, when Mitsuba doesn't find normals, will it try to calculate them automatically?

Edit: I just tried that and Mitsuba throws an error:

[xml.cpp:1079]   [OBJMesh] Error while loading OBJ file "subtractive-sphere-cut-meshed-5-nonormals.obj": reference to invalid normal 1!

That is because the "f" elements in the OBJ file need further correction, and my regex skills are rusty. So it would be helpful to know if this is worth it -- will Mitsuba calculate the normals if they are missing in the OBJ?

Thank you!


> Hello @buddhasystem,
> 
> Thanks for sharing your finding. I wonder if such high-resolution geometry is needed though, if the normals could be interpolated instead. Is there an option equivalent to [Blender's "Shade Smooth"](https://docs.blender.org/manual/en/2.91/scene_layout/object/editing/shading.html) in FreeCAD? Otherwise, you could try _not_ exporting the normals with your mesh and letting Mitsuba recalculate them. I think they will be smoothly interpolated by default (not sure).

**Reply by njroussel** [reply]
Hi @buddhasystem 

If the[ `face_normals` property](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_shapes.html#wavefront-obj-mesh-loader-obj) is set to `False` (default) and the file you loaded does not contain any normal information, Mitsuba will compute a smooth set of normals based on "Computing Vertex Normals from Polygonal Facets" by Grit Thuermer and Charles A. Wuethrich (JGT 1998, Vol 3)
